localstack/services/kinesis/kinesis_starter.py

```diff
@@ -21,9 +30,52 @@ def appy_patches():
 
 
 def start_kinesis(port=None, asynchronous=False, update_listener=None):
+    if KINESIS_PROVIDER == 'kinesis-mock':
+        return start_kinesis_mock(port=port, asynchronous=asynchronous, update_listener=update_listener)
+    if KINESIS_PROVIDER == 'kinesalite':
+        return start_kinesalite(port=port, asynchronous=asynchronous, update_listener=update_listener)
+    raise Exception('Unsupported Kinesis provider "%s"' % KINESIS_PROVIDER)
+
+
+def start_kinesis_mock(port=None, asynchronous=False, update_listener=None):
+    target_dir = os.path.join(INSTALL_DIR_INFRA, 'kinesis-mock')
+
+    if platform.startswith('linux'):
+        target_file_name = 'kinesis-mock-linux-amd64-dynamic'
+    elif platform == 'darwin':
+        target_file_name = 'kinesis-mock-macos-amd64-dynamic'
+    elif platform.startswith('win'):
+        target_file_name = 'kinesis-mock-dynamic.exe'
+    else:
+        target_file_name = 'kinesis-mock.jar'
+
+    target_file = os.path.join(target_dir, target_file_name)
+    if not os.path.exists(target_file):
+        response = requests.get(KINESIS_MOCK_RELEASES)
+        content = json.loads(to_str(response.content))
+        archive_url = content.get('assets', [])[0].get('browser_download_url')
+        download(archive_url, target_file)
+    port = port or config.PORT_KINESIS
+    backend_port = get_free_tcp_port()
+    kinesis_data_dir_param = ''
+    if config.DATA_DIR:
+        kinesis_data_dir = '%s/kinesis' % config.DATA_DIR
+        mkdir(kinesis_data_dir)
+        kinesis_data_dir_param = 'SHOULD_PERSIST_DATA=true PERSIST_PATH=%s' % kinesis_data_dir
+    if target_file_name.endswith('.jar'):
+        cmd = 'KINESIS_MOCK_HTTP1_PLAIN_PORT=%s SHARD_LIMIT=%s %s java -XX:+UseG1GC -jar %s' \
+            % (backend_port, config.KINESIS_SHARD_LIMIT, kinesis_data_dir_param, target_file)
+    else:
+        cmd = 'KINESIS_MOCK_HTTP1_PLAIN_PORT=%s SHARD_LIMIT=%s %s %s --gc=G1' \
```

Comment:
Looks awesome, thanks for taking over this PR @etspaceman !

Noticed a minor issue in the build:
```
/bin/sh: 1: /home/circleci/project/localstack/infra/kinesis-mock/kinesis-mock-linux-amd64-dynamic: Permission denied
```

The `chmod_r` function from `utils/common.py` should help:
```
chmod_r(target_file, 0o777)
```